File: elsalem_contracts_2025/elsalem_contracts_2025/overrides/material_request.py
"""
Material Request customizations for elsalem_contracts_2025
"""
import frappe
from frappe import _
from frappe.utils import nowdate


@frappe.whitelist()
def make_stock_entry_for_material_issue(source_name):
    """
    Create Stock Entry from Material Request with only items where request_type = 'Material Issue'
    
    Args:
        source_name: Name of the Material Request document
    
    Returns:
        Stock Entry document (dict)
    """
    # Get the Material Request document
    material_request = frappe.get_doc("Material Request", source_name)
    
    # Validate that the Material Request is submitted
    if material_request.docstatus != 1:
        frappe.throw(_("Material Request must be submitted before creating Stock Entry"))
    
    # Validate Material Request Type
    if material_request.material_request_type != "Issue & Purchase":
        frappe.throw(_("This function only works for Material Request Type 'Issue & Purchase'"))
    
    # Filter items where request_type = "Material Issue"
    filtered_items = [
        item for item in material_request.items 
        if item.get("request_type") == "Material Issue"
    ]
    
    # Check if there are any items to process
    if not filtered_items:
        frappe.msgprint(_("No items with 'Material Issue' request type found in this Material Request"))
        return None
    
    # Create new Stock Entry document
    stock_entry = frappe.new_doc("Stock Entry")
    stock_entry.stock_entry_type = "Material Issue"
    stock_entry.purpose = "Material Issue"
    stock_entry.company = material_request.company
    stock_entry.posting_date = nowdate()
    
    # Set warehouse from Material Request
    if material_request.set_warehouse:
        stock_entry.from_warehouse = material_request.set_warehouse
    
    # Transfer custom reason field if it exists
    if hasattr(material_request, "custom_reason") and material_request.custom_reason:
        stock_entry.custom_reason = material_request.custom_reason
    
    # Set Material Request reference
    stock_entry.material_request = material_request.name
    
    # Add filtered items to Stock Entry
    for mr_item in filtered_items:
        se_item = stock_entry.append("items", {})
        se_item.item_code = mr_item.item_code
        se_item.item_name = mr_item.item_name
        se_item.description = mr_item.description
        se_item.uom = mr_item.uom
        se_item.stock_uom = mr_item.stock_uom
        se_item.conversion_factor = mr_item.conversion_factor or 1
        se_item.qty = mr_item.qty
        
        # Set warehouse
        if material_request.set_warehouse:
            se_item.s_warehouse = material_request.set_warehouse
        
        # Set material request item reference
        se_item.material_request = material_request.name
        se_item.material_request_item = mr_item.name
        
        # Copy cost center and project if available
        if mr_item.cost_center:
            se_item.cost_center = mr_item.cost_center
        elif material_request.cost_center:
            se_item.cost_center = material_request.cost_center
            
        if mr_item.project:
            se_item.project = mr_item.project
    
    # Return the Stock Entry as dict
    return stock_entry.as_dict()

File: elsalem_contracts_2025/public/js/material_request.js
frappe.ui.form.on('Material Request', {
    // فلترة الأقسام حسب الشركة
    company: function (frm) {
        set_department_filter(frm);
    },
    // عند تحميل النموذج
    onload: function (frm) {
        set_department_filter(frm);
    },
    // عند تغيير القسم، جلب المدير
    department: function (frm) {
        if (frm.doc.department) {
            frappe.db.get_doc('Department', frm.doc.department).then(function (dept_doc) {
                if (dept_doc.expense_approvers && dept_doc.expense_approvers.length > 0) {
                    frm.set_value("manager", dept_doc.expense_approvers[0].approver);
                }
            });
        }
    },
    // الأزرار المخصصة عند الحالة المطلوبة
    refresh: function (frm) {
        // Set query filter for set_warehouse field
        frm.set_query('set_warehouse', () => {
            return {
                filters: {
                    is_group: 0  // Show only warehouses that are NOT groups
                }
            };
        });
        if (frm.doc.docstatus === 1 && frm.doc.material_request_type === "Issue & Purchase") {
            // 1. Stock Entry (Material Issue) - Only items with request_type = "Material Issue"
            frm.add_custom_button(__('Stock Entry (Material Issue)'), function () {
                frappe.call({
                    method: "elsalem_contracts_2025.elsalem_contracts_2025.overrides.material_request.make_stock_entry_for_material_issue",
                    args: {
                        source_name: frm.doc.name
                    },
                    callback: function (r) {
                        if (r.message) {
                            // Open the Stock Entry in a new form
                            frappe.model.sync(r.message);
                            frappe.set_route("Form", r.message.doctype, r.message.name);
                        }
                    }
                });
            }, __('Create'));
            // 2. Purchase Order
            frm.add_custom_button(__('Purchase Order'), function () {
                frappe.new_doc('Purchase Order', {
                    material_request: frm.doc.name,
                    company: frm.doc.company
                });
            }, __('Create'));
            // 3. Request for Quotation
            frm.add_custom_button(__('Request for Quotation'), function () {
                frappe.new_doc('Request for Quotation', {
                    material_request: frm.doc.name,
                    company: frm.doc.company
                });
            }, __('Create'));
            // 4. Supplier Quotation
            frm.add_custom_button(__('Supplier Quotation'), function () {
                frappe.new_doc('Supplier Quotation', {
                    material_request: frm.doc.name,
                    company: frm.doc.company
                });
            }, __('Create'));
        }
    }
});

// دالة فلترة الأقسام حسب الشركة
function set_department_filter(frm) {
    frm.set_query("department", function () {
        return {
            filters: {
                company: frm.doc.company
            }
        };
    });
}